frontend/src/components/dialogs/updateBook.tsx

```diff
@@ -0,0 +1,146 @@
+import React, { useEffect, useMemo, useRef, useState } from "react";
+import { Modal } from "antd";
+import { useMutation, useQuery } from "@apollo/client";
+import {
+  FormBuilder,
+  FormBuilderRef,
+  FormField,
+} from "@/components/formBuilder";
+import { GET_AUTHOR_NAMES } from "@/lib/api/authors";
+import {
+  GET_BOOK_LIST_QUERY,
+  GET_BOOKS_BY_AUTHOR_ID,
+  UPDATE_BOOK_MUTATION,
+} from "@/lib/api/books";
+import { useToast } from "@/hooks/use-toast";
+
+export interface UpdateBookProps {
+  open: boolean;
+  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
+  onComplete: () => void;
+  book: {
+    id: string;
+    title?: string;
+    description?: string;
+    publishedDate?: string;
+    authorId?: string;
+  };
+}
+
+export const UpdateBookDialog: React.FC<UpdateBookProps> = ({
+  open,
+  setOpen,
+  onComplete,
+  book,
+}) => {
+  const { toast } = useToast();
+  const formRef = useRef<FormBuilderRef>(null);
+  const updateError = useRef<any>(null);
+
+  const [mutateFunction, { data, loading, error }] = useMutation(
+    UPDATE_BOOK_MUTATION,
+    {
+      refetchQueries: [GET_BOOK_LIST_QUERY, GET_BOOKS_BY_AUTHOR_ID],
+    }
+  );
+
+  const { data: authorsData, refetch: refetchAuthors } = useQuery(
+    GET_AUTHOR_NAMES,
+    {
+      variables: {
+        pageSize: 10000,
+      },
+    }
+  );
+
+  const formDetails: FormField[] = useMemo(
+    () => [
+      {
+        type: "textarea",
+        name: "description",
+        label: "Description",
+        placeholder: "Enter new book description",
+      },
+      {
+        type: "date",
+        name: "publishedDate",
+        label: "Date of Publish",
+        placeholder: "Select date of publish",
+      },
+      {
+        type: "dropdown",
+        name: "authorId",
+        label: "Author",
+        placeholder: "Select author",
+        options: authorsData?.authors?.items?.map(
+          (item: { name: string; id: string }) => ({
+            label: item?.name ?? "",
+            value: item?.id ?? "",
+          })
+        ),
+      },
+    ],
+    [authorsData]
+  );
+
+  formRef.current?.reset();
+  refetchAuthors({
+    page: 1,
+    pageSize: 10000,
+  });
```

Comment:
Avoid calling functions directly inside component. This will be executed on all re-renders of `UpdateBookDialog`